resources/js/tableofcontent.js

```diff
@@ -1,3 +1,63 @@
+function isHeaderElement(elem) {
+    return elem.nodeName.match(/^H\d+$/) && elem.textContent;
+}
+
+function createHeaderData(elem, i) {
+    const headerText = elem.textContent.trim().replace(/"/g, '\\"');
+    const level = parseInt(elem.nodeName.substr(1));
+    const anchor = `kiwix-toc-${i}`;
+
+    const anchorElem = document.createElement("a");
+    anchorElem.id = anchor;
+
+    /* Mark header content with something we can reference. */
+    elem.insertAdjacentElement("afterbegin", anchorElem);
+
+    return { headerEntry: { text: headerText, anchor, table: [] }, level };
+}
+
+function getDOMHeadersPreorder(elem, array)
+{
+    if (isHeaderElement(elem))
+        array.push(elem);
+
+    Array.from(elem.children).forEach(child => getDOMHeadersPreorder(child, array));
+}
```

Comment:
For such name of the function (though I would make it slightly more generic `getDOMElementsPreorderDFS(elem, pred)`) I would prefer a clearer functional style implementation:

```js
function getDOMElementsPreorderDFS(elem, pred)
{
    var result = [];
    if (pred(elem))
        result.push(elem);

    for ( const child of elem.children) {
        result = result.concat(getDOMHeadersPreorder(child));
    }
    return result;
}
```

It may be a little slower because of the extra concatenations, however that should be typically negligible compared to the full iteration of the DOM.

If you want to keep the optimized version you better rename it to `collectHeaderElements(elem, outputArray)`.